DESIGN.md

```diff
@@ -0,0 +1,79 @@
+# RPC
+This project implements an RPC server in Go using gRPC as a transport.
+
+# Design
+What follows is a design document outlining the technology choices, architecture, and implementation of the RPC client and server. 
+
+## Authentication and Authorization
+The RPC service includes mutual TLS authentication. The client and server will both have their own private keys as well as certificates signed by the same CA. The minimum TLS version will be set to TLS1.2. In order to protect against downgrade attacks in TLS1.2, an allowlist will be used for ciphers. Conveniently, the Go TLS package will return a list of all cipher suites it implements excluding suites with known security issues (tls.CipherSuites()). In order to restrict the client to using these only, PreferServerCipherSuites will be set to true in the server TLS configuration. If this were a production application, an endpoint would be included in the API to accept a certificate revocation list, allowing the administrator to 'ban' certain clients. 
+
+For Authorization, a simple form of role based access control will be implemented: Each role will have its own CA. All CAs will be trusted by the server. Upon handling a remote procedure call, the server will first check the client certificate's certificate authority and compare it to its internal mapping of certificate authority to privilege level. The following roles will be implemented:
+- **Read only:** this role can get job output, but not create or delete jobs. This is useful for automated tooling such as status indicators and job notifiers.
+- **Read/Write access:** this role can get job output as well as create and delete jobs. This role is useful for administrators and any kind of deployment automation that would leverage this service.
+
+![visual representation of the multiple client CA model](doc/client_cas.png)
+
+## API
+### Endpoints
+The API exposed by the RPC service will contain the following endpoints
+- **CreateJob**: receives a Job and returns a CreateJobOutput
+- **DeleteJob**: receives a PID and returns a DeleteJobOutput
+- **ListJobs**: returns a JobList
+- **GetOutput**: receives a PID and returns a JobOutput
+
+### Datatypes
+- **Job**
+    - Command (string):
+    Command holds a valid shell command that is run on the underlying server shell.
+    - PID (integer):
+    PID holds an integer representing job ID. In this project it will correlate directly to the pid of the process running on the RPC server for a given job. In a production RPC system it is advisable to use a randomized PID.
+    - CreatedTime (string):
+    Timestamp that represents when the process was launched on the server
+- **JobList**
+    - Jobs ([]Job):
+    A list of Job obejcts as defined above in this list
+- **JobOutput**
+    - Command (string):
+    Holds the shell command as originally defined in the corresponding call to CreateJob
+    - CurrentOutput (string):
```

Comment:
If you're using grpc I'd suggest you create a separate endpoint for steaming the output instead of sending them as one chunk

Reply:
Streaming is a L3 req, and this design is currently targeting L2 (+grpc for convinence sake).  I'd avoid taking streaming scope unless you know you want to aim for L3.

Reply:
oh ok sounds good, I was thought we were targeting L3 now

Reply:
One more question. where will logs be stored? in memory or on disk?

Reply:
As I wrote in my latest comment on this PR, Jobs will be stored in memory.